### PR TITLE
OIDC Back-channel logout: Various intermediate enhancements

### PR DESCRIPTION
**Pull request by ayoho**
OP support added:
- Adds `sid` claim to all ID tokens the OP creates
- Adds `backchannelLogoutRequestTimeout` config attribute
    - Just added to the metatype - nothing is reading or using it yet
- Removes `backchannelLogoutSupported` config attribute
- Discovery:
    - In beta mode, will always return `backchannel_logout_supported` with a value of `true`
    - In beta mode, will always return `backchannel_logout_session_supported` with a value of `true`
- Framework laid for sending back-channel logout requests, but no real functionality 

RP supported added (OIDC and social):
- Adds `backchannelLogoutSupported` config attribute
- Adds `backchannelLogoutSessionRequired` config attribute
- Logout token validation:
    - Updated to require a `sid` claim in the logout token if `backchannelLogoutSessionRequired="true"`

### CODE REVIEW

**Comment by ayoho** [comment]
#libby

**Comment by ayoho** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.config,com.ibm.ws.security.oidc.client_fat.e2e,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.saml,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat,com.ibm.ws.security.oidc.server_fat.app-passwords,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-passwords.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.DerbyDB.opaque,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.jwt,com.ibm.ws.security.oidc.server_fat.app-tokens.LocalStore.opaque,com.ibm.ws.security.oidc.server_fat.clientregistration,com.ibm.ws.security.oidc.server_fat.config,com.ibm.ws.security.oidc.server_fat.config.oauth,com.ibm.ws.security.oidc.server_fat.config.oidc,com.ibm.ws.security.oidc.server_fat.endpoint.authorize,com.ibm.ws.security.oidc.server_fat.endpoint.coveragemap,com.ibm.ws.security.oidc.server_fat.endpoint.discovery,com.ibm.ws.security.oidc.server_fat.endpoint.end_session,com.ibm.ws.security.oidc.server_fat.endpoint.introspect,com.ibm.ws.security.oidc.server_fat.endpoint.revoke,com.ibm.ws.security.oidc.server_fat.endpoint.token,com.ibm.ws.security.oidc.server_fat.endpoint.userinfo,com.ibm.ws.security.oidc.server_fat.feature,com.ibm.ws.security.oidc.server_fat.granttype.jwt,com.ibm.ws.security.oidc.server_fat.gui.tokenMgmt,com.ibm.ws.security.oidc.server_fat.jaxrs,com.ibm.ws.security.oidc.server_fat.jaxrs.config,com.ibm.ws.security.oidc.server_fat.oauth,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.oidc.server_fat.saml,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2

**Comment by ayoho** [comment]
#libby

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops [hunk 1]
```diff
@@ -293,4 +293,9 @@ LOGOUT_TOKEN_EVENTS_MEMBER_VALUE_NOT_JSON=CWWKS1550E: The value for the "{0}" me
 LOGOUT_TOKEN_EVENTS_MEMBER_VALUE_NOT_JSON.explanation=The corresponding member value must be a JSON object. The expected value is "{}", an empty JSON object.
 LOGOUT_TOKEN_EVENTS_MEMBER_VALUE_NOT_JSON.useraction=Ensure that the "events" claim in the logout token is formatted correctly.
 
+LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING=CWWKS1551E: The logout token is missing the "sid" claim but the OpenID Connect client {0} requires it.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]:
```suggestion
LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING=CWWKS1551E: The logout token is missing the "sid" claim but the {0} OpenID Connect client requires it.
```

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops [hunk 2]
```diff
@@ -293,4 +293,9 @@ LOGOUT_TOKEN_EVENTS_MEMBER_VALUE_NOT_JSON=CWWKS1550E: The value for the "{0}" me
...
+LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING.explanation=The OpenID Connect client requires that the logout token must contain a session ID to identify the OpenID Connect client session with the OpenID Connect provider.
```
Comment (dmuelle) [inline dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]:
```suggestion
LOGOUT_TOKEN_SID_REQUIRED_BUT_MISSING.explanation=The OpenID Connect client requires that the logout token contains a session ID to identify the OpenID Connect client session with the OpenID Connect provider.
```

**Review by dmuelle — CHANGES_REQUESTED** [review]
ID review- two minor suggestions

**Comment by ayoho** [comment]
#libby

File: dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops (hunk 2 above)
Comment (ayoho) [reply dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/clients/common/resources/OidcClientMessages.nlsprops]: @dmuelle Just pushed a new commit

**Review by dmuelle — APPROVED** [review]
LGTM thanks

File: dev/com.ibm.ws.security.openidconnect.client/resources/OSGI-INF/metatype/metatype.xml [hunk 3]
```diff
@@ -154,6 +154,8 @@
          <AD id="keyManagementKeyAlias" name="%keyManagementKeyAlias" description="%keyManagementKeyAlias.desc" required="false" type="String" />
          <AD id="accessTokenCacheEnabled" name="%accessTokenCacheEnabled" description="%accessTokenCacheEnabled.desc" required="false" type="Boolean" default="true" />
          <AD id="accessTokenCacheTimeout" name="%accessTokenCacheTimeout" description="%accessTokenCacheTimeout.desc" required="false" type="String" default="5m" ibm:type="duration" />
+         <AD id="backchannelLogoutSupported" name="internal" description="internal use only" required="false" type="Boolean" default="false" ibm:beta="true" />
+         <AD id="backchannelLogoutSessionRequired" name="internal" description="internal use only" required="false" type="Boolean" default="false" ibm:beta="true" />
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.security.openidconnect.client/resources/OSGI-INF/metatype/metatype.xml:158]: To reflect current design, these should remain internal for GA.

File: dev/com.ibm.ws.security.social/resources/OSGI-INF/metatype/metatype.xml [hunk 4]
```diff
@@ -448,6 +448,8 @@
         <AD id="forwardLoginParameter" name="internal" description="internal use only" required="false" type="String" cardinality="2147483647" />
         <AD id="createSession" name="%createSession" description="%createSession.desc" required="false" type="Boolean" default="false" />
         <AD id="keyManagementKeyAlias" name="%keyManagementKeyAlias" description="%keyManagementKeyAlias.desc" required="false" type="String" />
+        <AD id="backchannelLogoutSupported" name="internal" description="internal use only" required="false" type="Boolean" default="false" ibm:beta="true" />
+        <AD id="backchannelLogoutSessionRequired" name="internal" description="internal use only" required="false" type="Boolean" default="false" ibm:beta="true" />
```
Comment (teddyjtorres) [inline dev/com.ibm.ws.security.social/resources/OSGI-INF/metatype/metatype.xml:452]: To reflect current design, these should remain internal for GA.

**Review by teddyjtorres — APPROVED** [review]

**Pull request merged**